Enable nested LoopByCount

annie-xd-wang opened the pull request:
This PR adds support for nested LoopByCount. The feature has been tested in synthetic mode using the following pipelines, and it is working as expected:

`[{"name": PrepareNextChannel,},((({"name": MoveToNextPositionInMultiPositionTable,"args": (None,None,None,),},{"name": ZStackAcquisition,"args": (False,False,"z-stack",False,),},{"name": WaitToContinue,},{"name": LoopByCount,"args": ("positions",False,),},),{"name": LoopByCount,"args": (3,False,),},),{"name": LoopByCount,"args": (4,False,),},),]`

annie-xd-wang committed:
update LoopByCount

annie-xd-wang committed:
typo

AdvancedImagingUTSW commented:
@conorhughmcfadden - You reported this error originally. Any chance you could evaluate it?

AdvancedImagingUTSW committed:
Update common_features.py

AdvancedImagingUTSW commented:
@annie-xd-wang - I updated the things that Copilot found which seemed accurate. If this works for @conorhughmcfadden we can proceed with merging.

conorhughmcfadden commented:
Unfortunately, I can't get Continuous Mode to run in this PR. I just boot up navigate and click Acquire > Continuous Mode and it hangs indefinitely with no image displayed, eventually I have to force quit and restart conda. Happens in synthetic mode as well which is very strange. 

Here is my most recent model_debug when trying this:

```console
2025-04-11 14:25:35,159 - model - INFO - decorators: DAQBase, (<DictProxy object, typeid 'dict' at 0x2cb5b0981f0>,), {}
2025-04-11 14:25:35,159 - model - INFO - decorators: SyntheticDAQ, (<DictProxy object, typeid 'dict' at 0x2cb5b0981f0>,), {}
2025-04-11 14:25:35,239 - model - INFO - decorators: MirrorBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>), {}
2025-04-11 14:25:35,239 - model - INFO - decorators: SyntheticMirror, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,299 - model - INFO - decorators: ZoomBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>), {}
2025-04-11 14:25:35,299 - model - INFO - decorators: SyntheticZoom, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,349 - model - INFO - decorators: ShutterBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>), {}
2025-04-11 14:25:35,349 - model - INFO - decorators: SyntheticShutter, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,419 - model - INFO - base: SyntheticCamera, Offset or variance map not found in C:\Users\vastopmv3\AppData\Local\.navigate\camera_maps
2025-04-11 14:25:35,419 - model - INFO - decorators: CameraBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>), {}
2025-04-11 14:25:35,419 - model - INFO - decorators: SyntheticCamera, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,529 - model - INFO - decorators: GalvoBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,529 - model - INFO - decorators: SyntheticGalvo, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,619 - model - INFO - decorators: FilterWheelBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,619 - model - INFO - decorators: SyntheticFilterWheel, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,719 - model - INFO - decorators: RemoteFocusBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>), {}
2025-04-11 14:25:35,719 - model - INFO - decorators: SyntheticRemoteFocus, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,789 - model - INFO - decorators: LaserBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,789 - model - INFO - decorators: SyntheticLaser, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,829 - model - INFO - decorators: LaserBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 1), {}
2025-04-11 14:25:35,829 - model - INFO - decorators: SyntheticLaser, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 1), {}
2025-04-11 14:25:35,949 - model - INFO - decorators: StageBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:35,949 - model - INFO - decorators: SyntheticStage, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 0), {}
2025-04-11 14:25:36,039 - model - INFO - decorators: StageBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 1), {}
2025-04-11 14:25:36,039 - model - INFO - decorators: SyntheticStage, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 1), {}
2025-04-11 14:25:36,139 - model - INFO - decorators: StageBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 2), {}
2025-04-11 14:25:36,139 - model - INFO - decorators: SyntheticStage, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 2), {}
2025-04-11 14:25:36,229 - model - INFO - decorators: StageBase, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 3), {}
2025-04-11 14:25:36,229 - model - INFO - decorators: SyntheticStage, ('OPM-VAST', None, <DictProxy object, typeid 'dict' at 0x2cb5b0981f0>, 3), {}
2025-04-11 14:25:37,529 - model - INFO - base: SyntheticCamera, Offset or variance map not found in C:\Users\vastopmv3\AppData\Local\.navigate\camera_maps
2025-04-11 14:25:38,579 - model - INFO - model: Change zoom of OPM-VAST to 1x
2025-04-11 14:25:38,689 - model - INFO - microscope: Microscope state: {'microscope_name': 'OPM-VAST', 'resolution_mode': 'Mesoscale', 'image_mode': 'z-stack', 'zoom': '1x', 'stack_cycling_mode': 'per_stack', 'start_position': -50.0, 'end_position': 50.0, 'step_size': 5.0, 'number_z_steps': 20, 'timepoints': 1, 'stack_pause': 0.0, 'etl_amplitude': 0, 'etl_offset': 0, 'is_save': False, 'stack_acq_time': 1.0, 'timepoint_interval': 0, 'experiment_duration': 1.05, 'is_multiposition': False, 'channels': <DictProxy object, typeid 'dict' at 0x2cb3ffb64f0>, 'stack_z_origin': -1, 'stack_focus_origin': -3, 'start_focus': 0.0, 'end_focus': 0.0, 'abs_z_start': -51.4, 'abs_z_end': 48.6, 'cycling': 'Per Stack', 'selected_channels': 1, 'scanrange': 100.0, 'n_plane': 1.0, 'offset_start': 0.0, 'offset_end': 0.0, 'conpro_cycling_mode': 'Per Plane', 'waveform_template': 'Default', 'multiposition_count': 3, 'autofocus_device': 'stage', 'autofocus_device_ref': 'f', 'primary_z_axis': 'z', 'primary_f_axis': 'f', 'secondary_stack_settings': <DictProxy object, typeid 'dict' at 0x2cb3ffb6880>}
2025-04-11 14:25:38,709 - model - INFO - microscope: Waveform constants: {'remote_focus_constants': <DictProxy object, typeid 'dict' at 0x2cb3ffb64f0>, 'galvo_constants': <DictProxy object, typeid 'dict' at 0x2cb3ffb6e50>, 'other_constants': <DictProxy object, typeid 'dict' at 0x2cb3ffb6670>}
2025-04-11 14:25:39,229 - model - INFO - microscope: Stage limits enabled: True
2025-04-11 14:25:47,983 - model - DEBUG - synthetic: Set camera trigger mode: External
2025-04-11 14:25:48,063 - model - DEBUG - microscope: Running microscope OPM-VAST
2025-04-11 14:25:48,083 - model - INFO - microscope: Preparing Acquisition. Camera Parameters: {'binning': '1x1', 'center_x': 1024, 'center_y': 1024, 'databuffer_size': 100, 'fov_x': 240, 'fov_y': 240, 'frames_to_average': 1.0, 'img_x_pixels': 1600, 'img_y_pixels': 1600, 'is_centered': True, 'number_of_pixels': 10, 'pixel_size': 6.5, 'readout_direction': 'Top-to-Bottom', 'readout_time': 10.0, 'sensor_mode': 'Normal', 'trigger_source': 'External', 'x_pixels': 1600, 'y_pixels': 1600}
2025-04-11 14:25:48,113 - model - INFO - microscope: Microscope state: {'microscope_name': 'OPM-VAST', 'resolution_mode': 'Mesoscale', 'image_mode': 'live', 'zoom': '1x', 'stack_cycling_mode': 'per_stack', 'start_position': -50.0, 'end_position': 50.0, 'step_size': 5.0, 'number_z_steps': 20, 'timepoints': 1, 'stack_pause': '0.0', 'etl_amplitude': 0, 'etl_offset': 0, 'is_save': False, 'stack_acq_time': 1.0, 'timepoint_interval': 0, 'experiment_duration': 1.05, 'is_multiposition': False, 'channels': <DictProxy object, typeid 'dict' at 0x2cb3ffb6130>, 'stack_z_origin': 0, 'stack_focus_origin': 0, 'start_focus': 0.0, 'end_focus': 0.0, 'abs_z_start': -50.0, 'abs_z_end': 50.0, 'cycling': 'Per Stack', 'selected_channels': 1, 'scanrange': 100.0, 'n_plane': 1.0, 'offset_start': 0.0, 'offset_end': 0.0, 'conpro_cycling_mode': 'Per Plane', 'waveform_template': 'Default', 'multiposition_count': 3, 'autofocus_device': 'stage', 'autofocus_device_ref': 'f', 'primary_z_axis': 'z', 'primary_f_axis': 'f', 'secondary_stack_settings': <DictProxy object, typeid 'dict' at 0x2cb3ffb6670>}
2025-04-11 14:25:48,123 - model - INFO - microscope: Waveform constants: {'remote_focus_constants': <DictProxy object, typeid 'dict' at 0x2cb3ffb6130>, 'galvo_constants': <DictProxy object, typeid 'dict' at 0x2cb3ffb6430>, 'other_constants': <DictProxy object, typeid 'dict' at 0x2cb3ffb64f0>}
2025-04-11 14:25:48,623 - model - INFO - feature_container: Running signal node: PrepareNextChannel
2025-04-11 14:25:48,663 - model - INFO - microscope: Camera exposure time set to 0.05.
2025-04-11 14:25:48,663 - model - INFO - microscope: 488 nm laser power set to 50.0
2025-04-11 14:25:48,703 - model - DEBUG - common_features: LoopByCount-initialize: 1, 1
2025-04-11 14:25:48,703 - model - DEBUG - common_features: LoopByCount-Synchronize signal
2025-04-11 14:25:53,856 - model - INFO - model: Running data process, getting frames []
2025-04-11 14:25:53,856 - model - DEBUG - model: Frame not received. Waiting 20/20 iterations
2025-04-11 14:25:58,927 - model - INFO - model: Running data process, getting frames []
2025-04-11 14:25:58,927 - model - DEBUG - model: Frame not received. Waiting 19/20 iterations
2025-04-11 14:26:03,999 - model - INFO - model: Running data process, getting frames []
2025-04-11 14:26:03,999 - model - DEBUG - model: Frame not received. Waiting 18/20 iterations
2025-04-11 14:26:03,999 - model - INFO - model: Data thread stopped.
2025-04-11 14:26:03,999 - model - INFO - model: Received frames in total: 0
2025-04-11 14:26:03,999 - model - INFO - microscope: Acquisition Ended
```

I checked and the model_debug is the same in synthetic or normal hardware mode. Seems the frames are never getting sent?

conorhughmcfadden commented:
> This PR adds support for nested LoopByCount. The feature has been tested in synthetic mode using the following pipelines, and it is working as expected:
> 
> `[{"name": PrepareNextChannel,},((({"name": MoveToNextPositionInMultiPositionTable,"args": (None,None,None,),},{"name": ZStackAcquisition,"args": (False,False,"z-stack",False,),},{"name": WaitToContinue,},{"name": LoopByCount,"args": ("positions",False,),},),{"name": LoopByCount,"args": (3,False,),},),{"name": LoopByCount,"args": (4,False,),},),]`

This does actually work in synthetic mode as @annie-xd-wang said, however. So the issue must be something specific to Continuous Mode?

annie-xd-wang committed:
update LoopByCount

annie-xd-wang commented:
Live mode works now. Here's the updated nested loop testing pipeline:

`[{"name": PrepareNextChannel,},((({"name": MoveToNextPositionInMultiPositionTable,"args": (None,None,None,),},{"name": ZStackAcquisition,"args": (False,False,"z-stack",False,),},{"name": WaitToContinue,},{"name": LoopByCount,"args": ("positions",False,True,),},),{"name": LoopByCount,"args": (3,False,True,),},),{"name": LoopByCount,"args": (4,False,False,),},),]`

AdvancedImagingUTSW commented:
I tested the continuous mode and it worked perfectly well on the BenchTop MesoSPIM. Hoping to have @conorhughmcfadden confirm that the LoopByCount works as he expects.

conorhughmcfadden committed:
Example working feature FlexibleLoopByParameter

- tested nested looping using my old feature which seemed to work
- using Fish Annotator to change multiposition_count on each iteration works as intended
- this is not the case with LoopByCount

conorhughmcfadden commented:
![image](https://github.com/user-attachments/assets/238dab58-81d7-44c4-95cb-8ef6bca12638)

So, the issue I'm having is, when using my Fish Annotator feature to change the number of positions on each iteration of an outer loop, self.signal does not update correctly in the inner loop. For example, running the feature list:

```console
[({"name": VastAnnotator, },{"name": PrepareNextChannel, },({"name": MoveToNextPositionInMultiPositionTable, },{"name": ZStackAcquisition, "args": (True,True,),},{"name": WaitToContinue, },{"name": LoopByCount, "args": ("positions",False,True,),},),{"name": LoopByCount, "args": (3,),},),]
```

... does not work, the number of positions remains the same each iteration regardless of how many points I select. [You can find the code for the VAST Annotator that I'm using here.](https://github.com/conorhughmcfadden/navigate-vast-interface/blob/main/navigate-vast-interface/controller/vast_interface_controller.py)

However, if I run using the feature I just pushed:

```console
[({"name": VastAnnotator,},{"name": PrepareNextChannel,},({"name": MoveToNextPositionInMultiPositionTable,},{"name": ZStackAcquisition,"args": (True,True,),},{"name": WaitToContinue,},{"name": FlexibleLoopByParameter,"args": ("experiment.MicroscopeState.multiposition_count",),},),{"name": LoopByCount,"args": (3,),},),]
```

It does have the correct number of positions on each loop, or at least self.signal is correct in initialize(). This is what I used to use several months ago and it seemed to work.

I made a few attempts to modify LoopByCount to do this, but found that the only way to get it to work was to make a new feature and remove the data thread, so it would only work with the signal thread (couldn't be used for Continuous Mode, for example).

There must be a way to do it so we don't have to add this extra looping feature, I just include it here for demonstration. 

The original problem was not that nested loops didn't work, it was that they didn't work _if_ the number of steps for the inner loop changed on each iteration of the outer loop. This is tough to test unless you're doing specifically what I'm doing here: selecting points on each iteration. But I imagine there would be a similar problem if you're e.g) segmenting different fields of view and wanting to high-res zoom in on your segments, etc.

conorhughmcfadden committed:
LoopByCount fix

- modifying self.steps on the first pass resulted in it being stored as int, so in later iterations get_steps() would not change the value
- instead we treat self.steps as a constant storing the input arg for steps. If it is a string, it gets reset each time get_steps() is called

The looping works as intended with my fish annotator plugin!

conorhughmcfadden committed:
Revert "Example working feature FlexibleLoopByParameter"

This reverts commit 58fbf7b5156998a3dc3988cf06cd56b01813b43e.

conorhughmcfadden commented:
@AdvancedImagingUTSW @annie-xd-wang  After some small changes with how `self.steps` is initialized, LoopByCount works as intended now. 

Tested and working with the following feature list:

```console
[({"name": VastAnnotator,},{"name": PrepareNextChannel,},({"name": MoveToNextPositionInMultiPositionTable,},{"name": ZStackAcquisition,"args": (True,True,),},{"name": WaitToContinue,},{"name": LoopByCount,"args": ("positions",False,True,),},),{"name": LoopByCount,"args": (3,False,False,),},),{"name": WaitToContinue,},]
```
